marker_driver_rustc/src/conversion/marker/pat.rs

```diff
@@ -0,0 +1,113 @@
+use marker_api::ast::pat::{
+    CommonPatData, IdentPat, OrPat, PatKind, RefPat, RestPat, SlicePat, StructFieldPat, StructPat, TuplePat,
+    UnstablePat, WildcardPat,
+};
+use rustc_hir as hir;
+
+use super::MarkerConversionContext;
+
+impl<'ast, 'tcx> MarkerConversionContext<'ast, 'tcx> {
+    #[must_use]
+    pub fn to_pat(&self, pat: &hir::Pat<'tcx>) -> PatKind<'ast> {
+        // Here we don't need to take special care for caching, as marker patterns
+        // don't have IDs and can't be requested individually. Instead patterns are
+        // stored as part of their parent expressions or items. Not needing to deal
+        // with caching makes this implementation simpler.
+        let data = CommonPatData::new(self.to_span_id(pat.span));
+
+        match &pat.kind {
+            hir::PatKind::Wild => PatKind::Wildcard(self.alloc(|| WildcardPat::new(data))),
+            hir::PatKind::Binding(hir::BindingAnnotation(by_ref, mutab), id, ident, pat) => {
+                PatKind::Ident(self.alloc(|| {
+                    IdentPat::new(
+                        data,
+                        self.to_symbol_id(ident.name),
+                        self.to_var_id(*id),
+                        matches!(mutab, rustc_ast::Mutability::Mut),
```

Comment:
Oh, also, shouldn't this be `hir::Mutability::Mut`?

Reply:
Rustc's HIR reuses these enums from `rustc_ast`. Most likely to make the conversion code simpler and since there is no new information added. The matches macro should also trigger a type mismatch error otherwise, since it's just expanded into a simple `match` statement.